File: jane_street/market_prediction/python/tests/test_engine.py
import numpy as np
import pytest
from market_prediction.engine import InferenceEngine
from market_prediction.models import ModelManager
from market_prediction.preprocessor import MarketPreprocessor

def test_inference_engine_predict_action(mock_model_manager, mock_feature_vector):
    preprocessor = MarketPreprocessor(n_features=130)
    engine = InferenceEngine(mock_model_manager, preprocessor, threshold=0.5)
    
    action = engine.predict_action(mock_feature_vector)
    assert action in [0, 1]

def test_inference_engine_batch_predict(mock_model_manager):
    preprocessor = MarketPreprocessor(n_features=130)
    engine = InferenceEngine(mock_model_manager, preprocessor, threshold=0.5)
    
    batch = np.random.randn(5, 130)
    actions = engine.batch_predict(batch)
    assert actions.shape == (5,)
    assert all(a in [0, 1] for a in actions)

File: jane_street/market_prediction/cpp/tests/test_engine.cpp
#include <gtest/gtest.h>
#include "market_prediction/engine.hpp"

using namespace market_prediction;

TEST(EngineTest, PredictAction) {
    auto model_manager = std::make_shared<ModelManager>(130);
    auto preprocessor = std::make_shared<MarketPreprocessor>(130);
    InferenceEngine engine(model_manager, preprocessor, 0.5);
    
    Eigen::VectorXd features = Eigen::VectorXd::Random(130);
    int action = engine.predict_action(features);
    EXPECT_TRUE(action == 0 || action == 1);
}

TEST(EngineTest, BatchPredict) {
    auto model_manager = std::make_shared<ModelManager>(130);
    auto preprocessor = std::make_shared<MarketPreprocessor>(130);
    InferenceEngine engine(model_manager, preprocessor, 0.5);
    
    Eigen::MatrixXd features_batch = Eigen::MatrixXd::Random(5, 130);
    auto actions = engine.batch_predict(features_batch);
    EXPECT_EQ(actions.size(), 5);
}
